Enhance Telnet option handling and viewport management

mystiker opened the pull request:
Introduce state management for Telnet options, including linemode and echo handling.

Implement viewport size management in the backend and update frontend components to reflect these changes.

Additionally, streamline logging and improve socket message handling.

mystiker committed:
experimental: new client buffer and linemode option handling

mystiker committed:
Merge remote-tracking branch 'origin/develop' into experimental/xterm-attach

mystiker committed:
chore: fixed cpy command

mystiker committed:
chore: sync packages

mystiker committed:
feat(telnet): enhance option handlers with state management and event listeners

- Updated `TelnetOptionHandler` to include generic state management methods: `getState`, `onStateChange`, and `offStateChange`.
- Refactored `handleEchoOption` to utilize a new `EchoNegotiator` class for managing echo state and event emissions.
- Introduced `LinemodeNegotiator` class in `handle-linemode-option` to manage linemode state and subnegotiation.
- Removed the `log-negotiation` utility and integrated logging directly into the `TelnetSocketWrapper`.
- Added linemode state management in the frontend, allowing for dynamic updates based on server negotiations.
- Created a new `LinemodeState` type to encapsulate linemode properties and updated the socket service to handle linemode events.
- Updated frontend components to reflect changes in local echo behavior based on linemode state.

mystiker committed:
feat(backend): implement viewport size management and connection handling

mystiker committed:
feat(frontend): enhance MudSocketAdapter with hooks for message transformation and lifecycle management

mystiker commented on frontend/src/app/features/sockets/types/server-to-client-events.ts:
```diff
@@ -1,7 +1,10 @@
+import { LinemodeState } from './linemode-state';
+
 export interface ServerToClientEvents {
```
Comment / Explain

mystiker replied:
Refactored into its own project, because it was used in both, frontend and backend.
Documented

mystiker commented on frontend/src/app/features/sockets/types/linemode-state.ts:
```diff
@@ -0,0 +1,9 @@
+export interface LinemodeState {
```
Comment / Explain

mystiker replied:
Refactored into its own project, because it was used in both, frontend and backend.
Documented

mystiker commented on frontend/src/app/features/sockets/types/client-to-server-events.ts:
```diff
@@ -1,5 +1,6 @@
 export interface ClientToServerEvents {
```
Comment / Explain

mystiker replied:
Refactored into its own project, because it was used in both, frontend and backend.
Documented

mystiker commented on backend/src/core/sockets/types/client-to-server-events.ts:
```diff
@@ -1,5 +1,6 @@
 export interface ClientToServerEvents {
```
Docs

mystiker replied:
Refactored into its own project, because it was used in both, frontend and backend.
Documented

mystiker commented on backend/src/core/sockets/types/server-to-client-events.ts:
```diff
@@ -1,7 +1,10 @@
 export interface ServerToClientEvents {
   mudOutput: (data: string) => void;
   mudDisconnected: () => void;
   mudConnected: () => void;
```
Docs

mystiker replied:
Refactored into its own project, because it was used in both, frontend and backend.
Documented

mystiker commented on frontend/src/app/core/mud/components/mud-client/mud-client.component.ts:
Wird in einem zukünftigen PR refactored

myonara reviewed (dismissed):
some doc comments open

mystiker committed:
feat(shared): add shared package with types and utilities

- Introduced a new shared package for common types and utilities used across frontend and backend.
- Added ServerConfig interface for backend configuration.
- Defined ClientToServerEvents and ServerToClientEvents interfaces for Socket.IO communication.
- Created LinemodeState interface to represent TELNET LINEMODE capabilities.
- Implemented TypeScript configuration for the shared package.
- Updated root package.json to include the shared workspace and adjusted build scripts.

mystiker committed:
fix(shared): removed need for existing scripts for workspaces
